File: labs/07-agent-evaluation/node/main.ts
import { OpenAIClient } from "../../../shared/config/openai_client";

const TEST_CASES = [
  {
    input: "What port range is safe for user services?",
    expected_keywords: ["1024", "65535"],
    reference: "User services should use ports from 1024 to 65535.",
  },
  {
    input: "Explain idempotency in one sentence.",
    expected_keywords: ["same", "multiple", "result"],
    reference:
      "Idempotency means calling an operation multiple times produces the same result.",
  },
];

function ruleCheck(answer: string, keywords: string[]): { passed: boolean; missing: string[] } {
  const missing = keywords.filter((kw) => !answer.toLowerCase().includes(kw.toLowerCase()));
  return { passed: missing.length === 0, missing };
}

async function llmJudge(client: OpenAIClient, answer: string, reference: string): Promise<any> {
  const prompt = `Rate how well the following answer matches the reference answer.
Answer: ${answer}
Reference: ${reference}
Respond with JSON only: {"score": 1-10, "reason": "..."}`;
  const response = await client.chatCompletion({
    messages: [{ role: "user", content: prompt }],
    response_format: { type: "json_object" },
    max_tokens: 200,
    temperature: 0.0,
  });
  return JSON.parse(response.choices[0].message.content);
}

async function evaluateAgent(
  client: OpenAIClient,
  agentFn: (client: OpenAIClient, input: string) => Promise<string>
): Promise<any[]> {
  const results: any[] = [];
  for (const testCase of TEST_CASES) {
    const answer = await agentFn(client, testCase.input);
    const ruleResult = ruleCheck(answer, testCase.expected_keywords);
    const judgeResult = await llmJudge(client, answer, testCase.reference);
    results.push({
      input: testCase.input,
      answer,
      rule_passed: ruleResult.passed,
      missing_keywords: ruleResult.missing,
      judge_score: judgeResult.score,
      judge_reason: judgeResult.reason,
    });
  }
  return results;
}

async function simpleAgent(client: OpenAIClient, question: string): Promise<string> {
  const response = await client.chatCompletion({
    messages: [{ role: "user", content: question }],
    max_tokens: 200,
    temperature: 0.0,
  });
  return response.choices[0].message.content;
}

async function main() {
  const client = new OpenAIClient();
  const results = await evaluateAgent(client, simpleAgent);

  const passed = results.filter((r) => r.rule_passed).length;
  const total = results.length;
  const avgScore = results.reduce((sum, r) => sum + (r.judge_score || 0), 0) / total;

  console.log(JSON.stringify(results, null, 2));
  console.log(`\nRule checks passed: ${passed}/${total}`);
  console.log(`Average judge score: ${avgScore.toFixed(1)}/10`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: labs/07-agent-evaluation/python/main.py
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "shared" / "config"))

from openai_client import OpenAIClient, configure_logging


TEST_CASES = [
    {
        "input": "What port range is safe for user services?",
        "expected_keywords": ["1024", "65535"],
        "reference": "User services should use ports from 1024 to 65535.",
    },
    {
        "input": "Explain idempotency in one sentence.",
        "expected_keywords": ["same", "multiple", "result"],
        "reference": "Idempotency means calling an operation multiple times produces the same result.",
    },
]


def rule_check(answer: str, keywords: list[str]) -> dict:
    missing = [kw for kw in keywords if kw.lower() not in answer.lower()]
    return {"passed": not missing, "missing": missing}


def llm_judge(client: OpenAIClient, answer: str, reference: str) -> dict:
    prompt = f"""Rate how well the following answer matches the reference answer.
Answer: {answer}
Reference: {reference}
Respond with JSON only: {{"score": 1-10, "reason": "..."}}"""
    response = client.chat_completion(
        messages=[{"role": "user", "content": prompt}],
        response_format={"type": "json_object"},
        max_tokens=200,
        temperature=0.0,
    )
    return json.loads(response["choices"][0]["message"]["content"])


def evaluate_agent(client: OpenAIClient, agent_fn) -> list[dict]:
    results = []
    for case in TEST_CASES:
        answer = agent_fn(client, case["input"])
        rule_result = rule_check(answer, case["expected_keywords"])
        judge_result = llm_judge(client, answer, case["reference"])
        results.append(
            {
                "input": case["input"],
                "answer": answer,
                "rule_passed": rule_result["passed"],
                "missing_keywords": rule_result["missing"],
                "judge_score": judge_result.get("score"),
                "judge_reason": judge_result.get("reason"),
            }
        )
    return results


def simple_agent(client: OpenAIClient, question: str) -> str:
    response = client.chat_completion(
        messages=[{"role": "user", "content": question}],
        max_tokens=200,
        temperature=0.0,
    )
    return response["choices"][0]["message"]["content"]


def main():
    configure_logging()
    client = OpenAIClient()
    results = evaluate_agent(client, simple_agent)

    passed = sum(1 for r in results if r["rule_passed"])
    total = len(results)
    avg_score = sum(r["judge_score"] or 0 for r in results) / total

    print(json.dumps(results, indent=2, ensure_ascii=False))
    print(f"\nRule checks passed: {passed}/{total}")
    print(f"Average judge score: {avg_score:.1f}/10")


if __name__ == "__main__":
    main()
